File: src/main/java/org/moinex/ui/main/SavingsBondsController.java
/*
 * Filename: SavingsBondsController.java
 * Created on: February 18, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.main;

import java.math.BigDecimal;
import java.text.MessageFormat;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javafx.beans.property.SimpleStringProperty;
import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.TextField;
import javafx.scene.text.Text;
import javafx.util.StringConverter;
import lombok.NoArgsConstructor;
import org.moinex.model.enums.BondType;
import org.moinex.model.investment.Bond;
import org.moinex.service.BondService;
import org.moinex.service.I18nService;
import org.moinex.ui.dialog.investment.AddBondController;
import org.moinex.ui.dialog.investment.AddBondPurchaseController;
import org.moinex.ui.dialog.investment.AddBondSaleController;
import org.moinex.ui.dialog.investment.ArchivedBondsController;
import org.moinex.ui.dialog.investment.BondTransactionsController;
import org.moinex.ui.dialog.investment.EditBondController;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public class SavingsBondsController {

    @FXML private Text bondsTabTotalInvestedField;
    @FXML private Text bondsTabCurrentValueField;
    @FXML private Text bondsTabProfitLossField;
    @FXML private Text bondsTabInterestReceivedField;
    @FXML private TableView<Bond> bondsTabBondTable;
    @FXML private TextField bondsTabBondSearchField;
    @FXML private ComboBox<BondType> bondsTabBondTypeComboBox;

    private ConfigurableApplicationContext springContext;
    private BondService bondService;
    private I18nService i18nService;

    private final Map<Integer, BigDecimal> currentMonthInterestCache = new HashMap<>();
    private final Map<Integer, BigDecimal> totalAccumulatedInterestCache = new HashMap<>();

    private static final Logger logger = LoggerFactory.getLogger(SavingsBondsController.class);

    @Autowired
    public SavingsBondsController(
            BondService bondService,
            ConfigurableApplicationContext springContext,
            I18nService i18nService) {
        this.bondService = bondService;
        this.springContext = springContext;
        this.i18nService = i18nService;
    }

    @FXML
    private void initialize() {
        configureBondTableView();
        updateBondTableView();
        updateBondTabFields();
        configureBondListeners();
    }

    @FXML
    private void handleRegisterBond() {
        WindowUtils.openModalWindow(
                Constants.ADD_BOND_FXML,
                i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_ADD_BOND_TITLE),
                springContext,
                (AddBondController controller) -> {},
                List.of(
                        () -> {
                            updateBondTableView();
                            updateBondTabFields();
                        }));
    }

    @FXML
    private void handleEditBond() {
        Bond selectedBond = bondsTabBondTable.getSelectionModel().getSelectedItem();

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .SAVINGS_BONDS_DIALOG_NO_SELECTION_EDIT_MESSAGE));
            return;
        }

        WindowUtils.openModalWindow(
                Constants.EDIT_BOND_FXML,
                i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_EDIT_BOND_TITLE),
                springContext,
                (EditBondController controller) -> controller.setBond(selectedBond),
                List.of(
                        () -> {
                            updateBondTableView();
                            updateBondTabFields();
                        }));
    }

    @FXML
    private void handleDeleteBond() {
        Bond selectedBond = bondsTabBondTable.getSelectionModel().getSelectedItem();

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .SAVINGS_BONDS_DIALOG_NO_SELECTION_DELETE_MESSAGE));
            return;
        }

        if (bondService.getOperationCountByBond(selectedBond.getId()) > 0) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_HAS_OPERATIONS_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_HAS_OPERATIONS_MESSAGE));
            return;
        }

        boolean confirmed =
                WindowUtils.showConfirmationDialog(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_DIALOG_CONFIRM_DELETE_TITLE),
                        MessageFormat.format(
                                i18nService.tr(
                                        Constants.TranslationKeys
                                                .SAVINGS_BONDS_DIALOG_CONFIRM_DELETE_MESSAGE),
                                selectedBond.getName()));

        if (confirmed) {
            try {
                bondService.deleteBond(selectedBond.getId());

                WindowUtils.showSuccessDialog(
                        i18nService.tr(
                                Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_BOND_DELETED_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_DIALOG_BOND_DELETED_MESSAGE));

                updateBondTableView();
                updateBondTabFields();
            } catch (IllegalStateException e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_DIALOG_ERROR_DELETING_BOND_TITLE),
                        e.getMessage());
            } catch (Exception e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(Constants.TranslationKeys.DIALOG_ERROR_TITLE),
                        e.getMessage());
            }
        }
    }

    @FXML
    private void handleOpenBondArchive() {
        WindowUtils.openModalWindow(
                Constants.ARCHIVED_BONDS_FXML,
                i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_BOND_ARCHIVE_TITLE),
                springContext,
                (ArchivedBondsController controller) -> {},
                List.of(
                        () -> {
                            updateBondTableView();
                            updateBondTabFields();
                        }));
    }

    @FXML
    private void handleBuyBond() {
        Bond selectedBond = bondsTabBondTable.getSelectionModel().getSelectedItem();

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .SAVINGS_BONDS_DIALOG_NO_SELECTION_BUY_MESSAGE));
            return;
        }

        WindowUtils.openModalWindow(
                Constants.BUY_BOND_FXML,
                i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_BUY_BOND_TITLE),
                springContext,
                (AddBondPurchaseController controller) -> {
                    controller.setBond(selectedBond);
                },
                List.of(
                        () -> {
                            updateBondTableView();
                            updateBondTabFields();
                        }));
    }

    @FXML
    private void handleSellBond() {
        Bond selectedBond = bondsTabBondTable.getSelectionModel().getSelectedItem();

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .SAVINGS_BONDS_DIALOG_NO_SELECTION_SELL_MESSAGE));
            return;
        }

        WindowUtils.openModalWindow(
                Constants.SALE_BOND_FXML,
                i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_DIALOG_SELL_BOND_TITLE),
                springContext,
                (AddBondSaleController controller) -> controller.setBond(selectedBond),
                List.of(
                        () -> {
                            updateBondTableView();
                            updateBondTabFields();
                        }));
    }

    @FXML
    private void handleShowBondTransactions() {
        WindowUtils.openModalWindow(
                Constants.BOND_TRANSACTIONS_FXML,
                i18nService.tr(Constants.TranslationKeys.BOND_DIALOG_TRANSACTIONS_TITLE),
                springContext,
                (BondTransactionsController controller) -> {},
                List.of(
                        () -> {
                            updateBondTableView();
                            updateBondTabFields();
                        }));
    }

    private void configureBondTableView() {
        TableColumn<Bond, String> nameColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_NAME));
        nameColumn.setCellValueFactory(
                cellData -> new SimpleStringProperty(cellData.getValue().getName()));

        TableColumn<Bond, String> symbolColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_SYMBOL));
        symbolColumn.setCellValueFactory(
                cellData -> new SimpleStringProperty(cellData.getValue().getSymbol()));

        TableColumn<Bond, String> typeColumn =
                new TableColumn<>(
                        i18nService.tr(Constants.TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_TYPE));
        typeColumn.setCellValueFactory(
                cellData ->
                        new SimpleStringProperty(
                                UIUtils.translateBondType(
                                        cellData.getValue().getType(), i18nService)));

        TableColumn<Bond, String> quantityColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_QUANTITY));
        quantityColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    return new SimpleStringProperty(
                            bondService.getCurrentQuantity(bond).toString());
                });

        TableColumn<Bond, String> avgPriceColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_TABLE_UNIT_PRICE));
        avgPriceColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    return new SimpleStringProperty(
                            UIUtils.formatCurrency(bondService.getAverageUnitPrice(bond)));
                });

        TableColumn<Bond, String> currentValueColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_TABLE_HEADER_CURRENT_VALUE));
        currentValueColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    return new SimpleStringProperty(
                            UIUtils.formatCurrency(bondService.getInvestedValue(bond).add(bondService.getTotalAccumulatedInterestByBondId(bond.getId()))));
                });

        TableColumn<Bond, String> investedValueColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_TABLE_HEADER_INVESTED_VALUE));
        investedValueColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    return new SimpleStringProperty(
                            UIUtils.formatCurrency(bondService.getInvestedValue(bond)));
                });

        TableColumn<Bond, String> maturityDateColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_MATURITY_DATE));
        maturityDateColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    if (bond.getMaturityDate() != null) {
                        return new SimpleStringProperty(
                                UIUtils.formatDateForDisplay(bond.getMaturityDate(), i18nService));
                    }
                    return new SimpleStringProperty("-");
                });

        TableColumn<Bond, String> interestRateColumn =
                new TableColumn<>(i18nService.tr(Constants.TranslationKeys.BOND_INTEREST_RATE));
        interestRateColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    if (bond.getInterestRate() != null) {
                        return new SimpleStringProperty(bond.getInterestRate() + "%");
                    }
                    return new SimpleStringProperty("-");
                });

        TableColumn<Bond, String> currentMonthInterestColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_TABLE_HEADER_CURRENT_MONTH_INTEREST));
        currentMonthInterestColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    BigDecimal interest =
                            currentMonthInterestCache.getOrDefault(bond.getId(), BigDecimal.ZERO);
                    return new SimpleStringProperty(UIUtils.formatCurrency(interest));
                });

        TableColumn<Bond, String> totalAccumulatedInterestColumn =
                new TableColumn<>(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .SAVINGS_BONDS_TABLE_HEADER_TOTAL_ACCUMULATED_INTEREST));
        totalAccumulatedInterestColumn.setCellValueFactory(
                cellData -> {
                    Bond bond = cellData.getValue();
                    BigDecimal interest =
                            totalAccumulatedInterestCache.getOrDefault(
                                    bond.getId(), BigDecimal.ZERO);
                    return new SimpleStringProperty(UIUtils.formatCurrency(interest));
                });

        bondsTabBondTable
                .getColumns()
                .addAll(
                        nameColumn,
                        symbolColumn,
                        typeColumn,
                        quantityColumn,
                        avgPriceColumn,
                        investedValueColumn,
                        currentValueColumn,
                        maturityDateColumn,
                        interestRateColumn,
                        currentMonthInterestColumn,
                        totalAccumulatedInterestColumn);
    }

    private void updateBondTableView() {
        List<Bond> bonds = bondService.getAllNonArchivedBonds();

        currentMonthInterestCache.clear();
        totalAccumulatedInterestCache.clear();
        for (Bond bond : bonds) {
            currentMonthInterestCache.put(
                    bond.getId(), bondService.getCurrentMonthInterest(bond.getId()));
            totalAccumulatedInterestCache.put(
                    bond.getId(), bondService.getTotalAccumulatedInterestByBondId(bond.getId()));
        }

        BondType selectedType = bondsTabBondTypeComboBox.getValue();
        if (selectedType != null) {
            bonds = bonds.stream().filter(bond -> bond.getType().equals(selectedType)).toList();
        }

        String searchText = bondsTabBondSearchField.getText().toLowerCase();
        if (!searchText.isEmpty()) {
            bonds =
                    bonds.stream()
                            .filter(
                                    bond -> {
                                        String name = bond.getName().toLowerCase();
                                        String symbol =
                                                bond.getSymbol() != null
                                                        ? bond.getSymbol().toLowerCase()
                                                        : "";
                                        String type =
                                                UIUtils.translateBondType(
                                                                bond.getType(), i18nService)
                                                        .toLowerCase();
                                        String issuer =
                                                bond.getIssuer() != null
                                                        ? bond.getIssuer().toLowerCase()
                                                        : "";
                                        String quantity =
                                                bondService.getCurrentQuantity(bond).toString();
                                        String avgPrice =
                                                bondService.getAverageUnitPrice(bond).toString();
                                        String investedValue =
                                                bondService.getInvestedValue(bond).toString();
                                        String profitLoss =
                                                bondService.calculateProfit(bond).toString();
                                        String maturityDate =
                                                bond.getMaturityDate() != null
                                                        ? UIUtils.formatDateForDisplay(
                                                                        bond.getMaturityDate(),
                                                                        i18nService)
                                                                .toLowerCase()
                                                        : "";
                                        String interestRate =
                                                bond.getInterestRate() != null
                                                        ? bond.getInterestRate().toString()
                                                        : "";

                                        return name.contains(searchText)
                                                || symbol.contains(searchText)
                                                || type.contains(searchText)
                                                || issuer.contains(searchText)
                                                || quantity.contains(searchText)
                                                || avgPrice.contains(searchText)
                                                || investedValue.contains(searchText)
                                                || profitLoss.contains(searchText)
                                                || maturityDate.contains(searchText)
                                                || interestRate.contains(searchText);
                                    })
                            .toList();
        }

        bondsTabBondTable.getItems().setAll(bonds);
    }

    private void updateBondTabFields() {
        List<Bond> bonds = bondService.getAllNonArchivedBonds();

        BigDecimal totalInvested = bondService.getTotalInvestedValue();

        BigDecimal profitLoss =
                bonds.stream()
                        .map(bondService::calculateProfit)
                        .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal interestReceived = bondService.getAllBondsTotalAccumulatedInterest();

        BigDecimal currentValue = totalInvested.add(interestReceived);

        bondsTabTotalInvestedField.setText(UIUtils.formatCurrency(totalInvested));
        bondsTabCurrentValueField.setText(UIUtils.formatCurrency(currentValue));
        bondsTabProfitLossField.setText(UIUtils.formatCurrencySigned(profitLoss));
        bondsTabInterestReceivedField.setText(UIUtils.formatCurrency(interestReceived));
    }

    private void configureBondListeners() {
        bondsTabBondTypeComboBox.setConverter(
                new StringConverter<>() {
                    @Override
                    public String toString(BondType bondType) {
                        if (bondType == null) {
                            return i18nService.tr(
                                    Constants.TranslationKeys.SAVINGS_STOCKS_FUNDS_FILTER_ALL);
                        }
                        return UIUtils.translateBondType(bondType, i18nService);
                    }

                    @Override
                    public BondType fromString(String string) {
                        return null;
                    }
                });

        bondsTabBondTypeComboBox.getItems().clear();
        bondsTabBondTypeComboBox.getItems().add(null);
        bondsTabBondTypeComboBox.getItems().addAll(BondType.values());
        bondsTabBondTypeComboBox.setValue(null);

        bondsTabBondTypeComboBox
                .valueProperty()
                .addListener((observable, oldValue, newValue) -> updateBondTableView());

        bondsTabBondSearchField
                .textProperty()
                .addListener((observable, oldValue, newValue) -> updateBondTableView());
    }
}

File: src/main/kotlin/org/moinex/ui/main/SavingsBondsController.kt
/*
 * Filename: SavingsBondsController.kt (original filename: SavingsBondsController.java)
 * Created on: February 18, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 17/03/2026
 */

package org.moinex.ui.main

import javafx.beans.property.SimpleStringProperty
import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.TextField
import javafx.scene.paint.Color
import javafx.scene.text.Text
import javafx.util.StringConverter
import org.moinex.common.constant.Files
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.enums.BondType
import org.moinex.model.investment.Bond
import org.moinex.service.PreferencesService
import org.moinex.service.investment.BondService
import org.moinex.ui.dialog.investment.create.AddBondController
import org.moinex.ui.dialog.investment.create.AddBondPurchaseController
import org.moinex.ui.dialog.investment.create.AddBondSaleController
import org.moinex.ui.dialog.investment.update.EditBondController
import org.moinex.ui.dialog.investment.view.ArchivedBondsController
import org.moinex.ui.dialog.investment.view.BondInterestHistoryController
import org.moinex.ui.dialog.investment.view.BondTransactionsController
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.text.MessageFormat

@Controller
class SavingsBondsController(
    private val bondService: BondService,
    private val springContext: ConfigurableApplicationContext,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var bondsTotalInvestedLabel: Text

    @FXML
    private lateinit var bondsTabTotalInvestedField: Text

    @FXML
    private lateinit var bondsCurrentValueLabel: Text

    @FXML
    private lateinit var bondsTabCurrentValueField: Text

    @FXML
    private lateinit var bondsProfitLossLabel: Text

    @FXML
    private lateinit var bondsTabProfitLossField: Text

    @FXML
    private lateinit var bondsInterestReceivedLabel: Text

    @FXML
    private lateinit var bondsTabInterestReceivedField: Text

    @FXML
    private lateinit var bondsTabBondTable: TableView<Bond>

    @FXML
    private lateinit var bondsTabBondSearchField: TextField

    @FXML
    private lateinit var bondsTabBondTypeComboBox: ComboBox<BondType>

    private val currentMonthInterestCache = mutableMapOf<Int, BigDecimal>()
    private val totalAccumulatedInterestCache = mutableMapOf<Int, BigDecimal>()

    @FXML
    fun initialize() {
        configureBondTableView()
        updateBondTableView()
        updateBondTabFields()
        configureBondListeners()
        applyThemeColors()
    }

    private fun applyThemeColors() {
        val color = if (preferencesService.isDarkMode()) Color.web("#F8F8F2") else Color.BLACK
        listOf(
            bondsTotalInvestedLabel,
            bondsTabTotalInvestedField,
            bondsCurrentValueLabel,
            bondsTabCurrentValueField,
            bondsProfitLossLabel,
            bondsTabProfitLossField,
            bondsInterestReceivedLabel,
            bondsTabInterestReceivedField,
        ).forEach { it.fill = color }
    }

    @FXML
    fun handleRegisterBond() {
        WindowUtils.openModalWindow(
            Files.ADD_BOND_FXML,
            preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_ADD_BOND_TITLE),
            springContext,
            { _: AddBondController -> },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    @FXML
    fun handleEditBond() {
        val selectedBond = bondsTabBondTable.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_EDIT_MESSAGE),
            )
            return
        }

        WindowUtils.openModalWindow(
            Files.EDIT_BOND_FXML,
            preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_EDIT_BOND_TITLE),
            springContext,
            { controller: EditBondController -> controller.setBond(selectedBond) },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    @FXML
    fun handleDeleteBond() {
        val selectedBond = bondsTabBondTable.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_DELETE_MESSAGE),
            )
            return
        }

        if (bondService.getOperationCountByBond(selectedBond.id!!) > 0) {
            WindowUtils.showErrorDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_HAS_OPERATIONS_TITLE),
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_HAS_OPERATIONS_MESSAGE),
            )
            return
        }

        val confirmed =
            WindowUtils.showConfirmationDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_CONFIRM_DELETE_TITLE),
                MessageFormat.format(
                    preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_CONFIRM_DELETE_MESSAGE),
                    selectedBond.name,
                ),
            )

        if (confirmed) {
            runCatching {
                bondService.deleteBond(selectedBond.id!!)

                WindowUtils.showSuccessDialog(
                    preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_BOND_DELETED_TITLE),
                    preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_BOND_DELETED_MESSAGE),
                )

                updateBondTableView()
                updateBondTabFields()
            }.onFailure { e ->
                val message =
                    when (e) {
                        is IllegalStateException ->
                            preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_ERROR_DELETING_BOND_TITLE)

                        else -> preferencesService.translate(TranslationKeys.DIALOG_ERROR_TITLE)
                    }
                WindowUtils.showErrorDialog(message, e.message ?: "")
            }
        }
    }

    @FXML
    fun handleOpenBondArchive() {
        WindowUtils.openModalWindow(
            Files.ARCHIVED_BONDS_FXML,
            preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_BOND_ARCHIVE_TITLE),
            springContext,
            { _: ArchivedBondsController -> },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    @FXML
    fun handleBuyBond() {
        val selectedBond = bondsTabBondTable.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_BUY_MESSAGE),
            )
            return
        }

        WindowUtils.openModalWindow(
            Files.BUY_BOND_FXML,
            preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_BUY_BOND_TITLE),
            springContext,
            { controller: AddBondPurchaseController -> controller.initializeBond(selectedBond) },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    @FXML
    fun handleSellBond() {
        val selectedBond = bondsTabBondTable.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_SELL_MESSAGE),
            )
            return
        }

        WindowUtils.openModalWindow(
            Files.SALE_BOND_FXML,
            preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_SELL_BOND_TITLE),
            springContext,
            { controller: AddBondSaleController -> controller.initializeBond(selectedBond) },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    @FXML
    fun handleShowBondTransactions() {
        WindowUtils.openModalWindow(
            Files.BOND_TRANSACTIONS_FXML,
            preferencesService.translate(TranslationKeys.BOND_DIALOG_TRANSACTIONS_TITLE),
            springContext,
            { _: BondTransactionsController -> },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    @FXML
    fun handleShowInterestHistory() {
        val selectedBond = bondsTabBondTable.selectionModel.selectedItem

        if (selectedBond == null) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_TITLE),
                preferencesService.translate(TranslationKeys.SAVINGS_BONDS_DIALOG_NO_SELECTION_HISTORY_MESSAGE),
            )
            return
        }

        WindowUtils.openModalWindow(
            Files.BOND_INTEREST_HISTORY_FXML,
            MessageFormat.format(
                preferencesService.translate(TranslationKeys.BOND_DIALOG_INTEREST_HISTORY_TITLE),
                selectedBond.name,
            ),
            springContext,
            { controller: BondInterestHistoryController -> controller.setBond(selectedBond) },
            listOf(
                Runnable {
                    updateBondTableView()
                    updateBondTabFields()
                },
            ),
        )
    }

    private fun configureBondTableView() {
        bondsTabBondTable.columns.addAll(
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_NAME,
            ) { it.name },
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_SYMBOL,
            ) { it.symbol ?: "" },
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_TYPE,
            ) { UIUtils.translateBondType(it.type) },
            createColumn(
                TranslationKeys.BOND_TABLE_QUANTITY,
            ) { bondService.getCurrentQuantity(it).toString() },
            createColumn(
                TranslationKeys.BOND_TABLE_UNIT_PRICE,
            ) { UIUtils.formatCurrency(bondService.getAverageUnitPrice(it)) },
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_INVESTED_VALUE,
            ) { UIUtils.formatCurrency(bondService.getTotalInvestedValue(it)) },
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_CURRENT_VALUE,
            ) {
                UIUtils.formatCurrency(
                    bondService.getTotalInvestedValue(it).add(
                        bondService.getTotalAccumulatedInterestByBondId(it.id!!),
                    ),
                )
            },
            createColumn(
                TranslationKeys.BOND_MATURITY_DATE,
            ) { it.maturityDate?.let { date -> UIUtils.formatDateForDisplay(date) } ?: "-" },
            createColumn(
                TranslationKeys.BOND_INTEREST_RATE,
            ) { it.interestRate?.let { rate -> "$rate%" } ?: "-" },
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_CURRENT_MONTH_INTEREST,
            ) {
                UIUtils.formatCurrency(
                    currentMonthInterestCache.getOrDefault(it.id, BigDecimal.ZERO),
                )
            },
            createColumn(
                TranslationKeys.SAVINGS_BONDS_TABLE_HEADER_TOTAL_ACCUMULATED_INTEREST,
            ) {
                UIUtils.formatCurrency(
                    totalAccumulatedInterestCache.getOrDefault(it.id, BigDecimal.ZERO),
                )
            },
        )
    }

    private fun createColumn(
        headerKey: String,
        valueExtractor: (Bond) -> String,
    ): TableColumn<Bond, String> =
        TableColumn<Bond, String>(preferencesService.translate(headerKey)).apply {
            setCellValueFactory { SimpleStringProperty(valueExtractor(it.value)) }
        }

    private fun updateBondTableView() {
        val bonds = bondService.getAllNonArchivedBonds()

        currentMonthInterestCache.clear()
        totalAccumulatedInterestCache.clear()

        bonds.forEach { bond ->
            currentMonthInterestCache[bond.id!!] = bondService.getCurrentMonthInterest(bond.id!!)
            totalAccumulatedInterestCache[bond.id!!] = bondService.getTotalAccumulatedInterestByBondId(bond.id!!)
        }

        val filteredBonds =
            bonds
                .filter { bond ->
                    bondsTabBondTypeComboBox.value?.let { bond.type == it } ?: true
                }.filter { bond ->
                    val searchText = bondsTabBondSearchField.text.lowercase()
                    if (searchText.isEmpty()) {
                        true
                    } else {
                        listOf(
                            bond.name.lowercase(),
                            bond.symbol?.lowercase() ?: "",
                            UIUtils.translateBondType(bond.type).lowercase(),
                            bond.issuer?.lowercase() ?: "",
                            bondService.getCurrentQuantity(bond).toString(),
                            bondService.getAverageUnitPrice(bond).toString(),
                            bondService.getTotalInvestedValue(bond).toString(),
                            bondService.getTotalProfit(bond).toString(),
                            bond.maturityDate?.let {
                                UIUtils.formatDateForDisplay(it).lowercase()
                            } ?: "",
                            bond.interestRate?.toString() ?: "",
                        ).any { it.contains(searchText) }
                    }
                }

        bondsTabBondTable.items.setAll(filteredBonds)
    }

    private fun updateBondTabFields() {
        val bonds = bondService.getAllNonArchivedBonds()

        val totalInvested = bondService.getTotalInvestedValue()

        val profitLoss =
            bonds
                .map { bondService.getTotalProfit(it) }
                .fold(BigDecimal.ZERO, BigDecimal::add)

        val interestReceived = bondService.getAllBondsTotalAccumulatedInterest()

        val currentValue = totalInvested.add(interestReceived)

        bondsTabTotalInvestedField.text = UIUtils.formatCurrency(totalInvested)
        bondsTabCurrentValueField.text = UIUtils.formatCurrency(currentValue)
        bondsTabProfitLossField.text = UIUtils.formatCurrencySigned(profitLoss)
        bondsTabInterestReceivedField.text = UIUtils.formatCurrency(interestReceived)
    }

    private fun configureBondListeners() {
        bondsTabBondTypeComboBox.converter =
            object : StringConverter<BondType>() {
                override fun toString(bondType: BondType?): String =
                    bondType?.let {
                        UIUtils.translateBondType(it)
                    } ?: preferencesService.translate(TranslationKeys.SAVINGS_STOCKS_FUNDS_FILTER_ALL)

                override fun fromString(string: String): BondType? = null
            }

        bondsTabBondTypeComboBox.items.apply {
            clear()
            add(null)
            addAll(BondType.entries)
        }
        bondsTabBondTypeComboBox.value = null

        bondsTabBondTypeComboBox.valueProperty().addListener { _, _, _ -> updateBondTableView() }
        bondsTabBondSearchField.textProperty().addListener { _, _, _ -> updateBondTableView() }
    }
}
